sdk/search/azure-search/src/main/java/com/azure/search/SearchIndexClientBuilder.java

```diff
@@ -0,0 +1,265 @@
+ * {@link SearchIndexClient SearchIndexClients} and {@link SearchIndexAsyncClient SearchIndexAsyncClients}.
+ * Call {@link #buildClient() buildClient} and {@link #buildAsyncClient() buildAsyncClient} respectively to construct
+ * an instance of the desired client.
+ *
+ * <p>
+ * The following information must be provided on this builder:
+ *     <ul>
+ *         <li>the Azure Cognitive Search service endpoint through {@code .endpoint()}
+ *         <li>the index name through {@code .indexName()}
+ *         <li>the API key through {@code .credential()}</li>
+ *     </ul>
+ * </p>
+ */
+@ServiceClientBuilder(serviceClients = {SearchIndexClient.class, SearchIndexAsyncClient.class})
+public class SearchIndexClientBuilder {
+
+    // This header tells the server to return the request id in the HTTP response. Useful for correlation with what
+    // request was sent.
+    private static final String ECHO_REQUEST_ID_HEADER = "x-ms-return-client-request-id";
+
+    private static final String SEARCH_PROPERTIES = "azure-search.properties";
+    private static final String NAME = "name";
+    private static final String VERSION = "version";
+
+    SearchApiKeyCredential searchApiKeyCredential;
+    SearchServiceVersion apiVersion;
+    String endpoint;
+    HttpClient httpClient;
+    HttpLogOptions httpLogOptions;
+    Configuration configuration;
+    List<HttpPipelinePolicy> policies;
+    private String clientName;
+    private String clientVersion;
+
+    private String indexName;
+    private final HttpHeaders headers;
+    private RetryPolicy retryPolicy;
+    private final ClientLogger logger = new ClientLogger(SearchIndexClientBuilder.class);
+
+    /**
+     * Default Constructor
+     */
+    public SearchIndexClientBuilder() {
+        apiVersion = SearchServiceVersion.getLatest();
+        policies = new ArrayList<>();
+        httpClient = HttpClient.createDefault();
+        httpLogOptions = new HttpLogOptions();
+
+        Map<String, String> properties = CoreUtils.getProperties(SEARCH_PROPERTIES);
+        clientName = properties.getOrDefault(NAME, "UnknownName");
+        clientVersion = properties.getOrDefault(VERSION, "UnknownVersion");
+        headers = new HttpHeaders()
+            .put(ECHO_REQUEST_ID_HEADER, "true");
+    }
+
+    /**
+     * Sets the api version to work against
+     *
+     * @param apiVersion api version
+     * @return the updated SearchIndexClientBuilder object
+     */
+    public SearchIndexClientBuilder apiVersion(SearchServiceVersion apiVersion) {
+        if (apiVersion == null) {
+            throw logger.logExceptionAsError(new IllegalArgumentException("Invalid apiVersion"));
+        }
+        this.apiVersion = apiVersion;
+        return this;
+    }
+
+    /**
+     *  Returns the list of policies configured for this builder.
+     *
+     *  @return List of HttpPipelinePolicy
+     *  */
+    public List<HttpPipelinePolicy> getPolicies() {
```

Comment:
We don't usually have public getter methods in Builder class. Is there a reason for having this getter?

Reply:
used in the builder's unit tests

Reply:
If it's required for unit tests, can it be just package-private instead of public?